tools/samtools.py

```diff
@@ -182,6 +182,59 @@ def removeDoublyMappedReads(self, inBam, outBam):
         opts = ['-b', '-F' '1028', '-f', '2', '-@', '3']
         self.view(opts, inBam, outBam)
 
+    def filter_to_mapped_reads(self, inBam, outBam, allow_unmapped=True, min_mapping_qual=None, remove_singletons=True):
+        '''
+            This function writes a bam file filtered to include properly aligned reads.
+            If allow_unmapped=True, fully-unmapped pairs or unmapped single-end reads are also 
+            written to the output (omitting pairs where only one mate maps).
+
+        '''
+
+        with pysam.AlignmentFile(inBam, 'rb', check_sq=False) as inb:
+            with pysam.AlignmentFile(outBam, 'wb', header=inb.header) as outf:
+                # process the lines individually and write them or not, depending on the flags
+                # For explanation of flags, see:
+                # https://broadinstitute.github.io/picard/explain-flags.html
+                # https://pysam.readthedocs.io/en/latest/api.html
+                # https://samtools.github.io/hts-specs/SAMv1.pdf
+                # https://github.com/pysam-developers/pysam/blob/31183d7fac52b529b304bdf61ff933818ae4a71f/samtools/stats.c#L72-L81
+
+                for read in inb:
+                    # check if a read is paired
+                    is_single_end=not read.is_paired
```

Comment:
wait, is this what you want? I think this boolean tells you whether the read came off the sequencer as part of a paired-end sequencing run, not whether both mates are aligned.

Reply:
Yeah, the intent here was to distinguish whether mapped reads are single-end (and allowed through) or (paired) singletons (where they're rejected).
Maybe we don't worry about this function having an option to allow unmapped, and simplify to something like:
```
                    # if not a properly-mapped pair (however an alignment defines that)
                    # or not a mapped single-end read
                    # do not write
                    if not read.is_proper_pair or (read.is_unmapped and not read.is_paired):
                        continue 
                    # if a PCR/optical duplicate, do not write
                    if read.is_duplicate:
                        continue
```
With a mapping quality filter immediately upstream.

I'm unclear on whether there are circumstances where the `is_proper_pair` bit may be set for secondary alignments or alignments that don't really make sense like supplementary (chimeric) alignments, but aligner behavior does in general differ in the criteria used to flag proper pairing. We can probably toss secondary alignments since there should be a primary somewhere. Not sure about supplementary alignments; if a read is mapped but its mate only has a supplementary alignment, do we keep it? (or both?) Would any aligner consider that proper pairing? We couldn't reject both in a streaming/iterator way since there doesn't seem to be a flag to know what type of alignment a read's mate has beyond whether it is mapped  at all and whether it is mapped to the reverse strand.